tests/xharness/Harness.cs

```diff
@@ -80,52 +106,80 @@ public class Harness : IHarness
+		public string MAC_DESTDIR { get; private set; }
+		public string IOS_DESTDIR { get; private set; }
+		public string MONO_IOS_SDK_DESTDIR { get; private set; }
+		public string MONO_MAC_SDK_DESTDIR { get; private set; }
+		public bool IncludeMac32 { get; private set; }
+		public bool ENABLE_XAMARIN { get; private set; }
 
 		// Run
-		public AppRunnerTarget Target { get; set; }
-		public string SdkRoot { get; set; }
-		public string Configuration { get; set; } = "Debug";
-		public string LogFile { get; set; }
-		public string LogDirectory { get; set; } = Environment.CurrentDirectory;
-		public double Timeout { get; set; } = 15; // in minutes
-		public double LaunchTimeout { get; set; } // in minutes
-		public bool DryRun { get; set; } // Most things don't support this. If you need it somewhere, implement it!
-		public string JenkinsConfiguration { get; set; }
-		public Dictionary<string, string> EnvironmentVariables { get; set; } = new Dictionary<string, string> ();
-		public string MarkdownSummaryPath { get; set; }
-		public string PeriodicCommand { get; set; }
-		public string PeriodicCommandArguments { get; set; }
-		public TimeSpan PeriodicCommandInterval { get; set; }
+		public AppRunnerTarget Target { get; }
+		public string SdkRoot { get; private set; }
+		public string Configuration { get; }
+		public string LogFile { get; }
+		public string LogDirectory { get; }
+		public double Timeout { get; } = 15; // in minutes
+		public double LaunchTimeout { get; } // in minutes
+		public bool DryRun { get; } // Most things don't support this. If you need it somewhere, implement it!
+		public string JenkinsConfiguration { get; }
+		public Dictionary<string, string> EnvironmentVariables { get; }
+		public string MarkdownSummaryPath { get; }
+		public string PeriodicCommand { get; }
+		public string PeriodicCommandArguments { get; }
+		public TimeSpan PeriodicCommandInterval { get; }
 		// whether tests that require access to system resources (system contacts, photo library, etc) should be executed or not
 		public bool? IncludeSystemPermissionTests { get; set; }
 
-		public Harness ()
-		{
+		public Harness (IProcessManager processManager, HarnessAction action, HarnessConfiguration configuration)
+		{
+			ProcessManager = processManager ?? throw new ArgumentNullException (nameof (processManager));
+			Action = action;
+
+			AutoConf = configuration.AutoConf;
+			Configuration = configuration.Configuration ?? throw new ArgumentNullException (nameof (configuration));
+			DryRun = configuration.DryRun;
+			EnvironmentVariables = configuration.EnvironmentVariables ?? new Dictionary<string, string> ();
+			IncludeSystemPermissionTests = configuration.IncludeSystemPermissionTests;
+			IOSTestProjects = configuration.IOSTestProjects;
+			JenkinsConfiguration = configuration.JenkinsConfiguration;
+			Labels = configuration.Labels ?? new HashSet<string> ();
```

Comment:
I worry about this - the `Labels` property is set to the object in the configuration. If the configuration changes, so does this since they share the same object. Maybe it would be better to set the `Labels` to a default empty set and use `if (configuration.Labels != null) Labels.UnionWith (configuration.Labels);` I see the same issue with `EnvironmentVariables` as well.

Reply:
I am very much against having the `HarnessConfiguration` wrapper because it is misleading on what it does when we change it.

I was considering referencing the `HarnessConfiguration` inside `Harness` and then have the other properties point to it.

But to be honest, I would just not have it at all and have everything handed over in ctor like on the previous iteration of this PR.

Reply:
But a good point otherwise